Add support for component-installer v3

boesing opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| BC Break      | no
| New Feature   | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

We do use internal stuff from our composer plugin here. Not sure if we should keep it like this but for the moment, this works by allowing the access of the internal classes from the component installer.

Maybe we should extract the injection functionality to a dedicated component and use it here and within the plugin.

Ref: https://github.com/laminas/laminas.dev/pull/24

boesing committed:
feature: add support for component-installer v3

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented on src/Module/DeregisterCommand.php:
```diff
@@ -68,8 +69,15 @@ protected function execute(InputInterface $input, OutputInterface $output): int
         $module   = $input->getArgument('module');
         $composer = $input->getOption('composer') ?: 'composer';
 
+        /**
+         * @psalm-suppress InternalClass,InternalMethod We do explicitly need the functionality of the component
+         *                                              installer here
+         */
```
This will break though, since we can remove this class at any time in upstream (if it's `@internal`) :thinking:

boesing replied:
Yah, but I think we can go with this for now since mezzio-tooling and the component-installer are development-only deps and therefore should not break production code.

Not sure if we should address this for now. WDYT?

Ocramius replied:
> are development-only deps and therefore should not break production code.

They will still break **within** the organization, which means more maintenance work for us internally.

boesing replied:
Copy & pasting stuff also creates more maintenance work?
If I now have to copy & paste stuff here, that will take some time as well.

I would rather create a dedicated component which handles the config injection.
Currently, that only works with `regex` and I would love to use `nikic/php-parser` for injection instead.
So if we want to duplicate the logic here, I would prefer extracting all the configuration manipulation logic to a dedicated component.

Ocramius replied:
> Copy & pasting stuff also creates more maintenance work?

It certainly doesn't pull the rug under our feet, heh :D 

Planned maintenance is better than unplanned one, no?

boesing replied:
Will copy & paste the logic into this component. extracting it to a dedicated component can still be achieved if we find a bug either in the composer plugin or in this component.

Mantra: Fix problems when there are problems.

Ocramius reviewed (approved):
Besides the `@internal` usages, LGTM.

I wonder if copy-pasting those classes would be better...

boesing committed:
qa: import internal classes from component-installer into this package

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented on src/ConfigDiscovery/AbstractDiscovery.php:
```diff
@@ -0,0 +1,59 @@
+/**
+ * @internal
+ */
+abstract class AbstractDiscovery implements DiscoveryInterface
```
Tests for these symbols should also be imported 🤔

Ocramius replied:
Sorry, I totally missed that in my previous review :(

boesing replied:
We do actually test it by integration testing the whole component (e.g. that the `ConfigProvider` is injected to the appropriate file on the FS). 

Do you still want me to unit-test these files even if its covered by integration tests?
just give me a quick thumbs up if so, if not, thumbs down.

Ocramius replied:
Mostly just copying over the tests from the other components would suffice.

Reasoning is that as we apply CS fixes and static-analysis related changes/refactoring, if we have coverage, we're OK. Very risky otherwise.

weierophinney replied:
Done!

weierophinney committed:
qa: adds tests for classes imported from laminas-component-installer

- `ConfigDiscovery\ConfigAggregatorTest`
- `ConfigInjector\AbstractInjectorTestCase`
- `ConfigInjector\ConfigAggregatorInjectorTest`
- related test assets

Additionally, the Register/Deregister command tests needed to mark several test methods to run in separate processes and ignore global state to ensure that class overriding would work.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: resolve errors flagged by psalm

- Marks internal classes and methods for suppression
- Identifies one `ArgumentTypeCoercion` in a test that is valid, and adds to the baseline.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
@boesing I've imported the tests and pushed them to your branch.

Ocramius commented on src/Exception/RuntimeException.php:
```diff
@@ -0,0 +1,11 @@
+
+use RuntimeException as PhpRuntimeException;
+
+final class RuntimeException extends PhpRuntimeException
```
Probably needs `@internal`, but I doubt anybody will use this.

Ocramius reviewed (approved):
:ship: thanks @boesing @weierophinney!

boesing commented:
Thanks, quite busy ATM. Wanted to work on this next week but thanks!
